cristin-project/src/main/java/no/unit/nva/cristin/projects/common/CristinQueryBuilder.java

```diff
@@ -0,0 +1,613 @@
+        cristinQuery.isNvaQuery = true;
+        var nonNullValue = nonNull(value) ? value : EMPTY_STRING;
+
+        if (key.equals(IDENTITY.getNvaKey())) {
+            withPathIdentity(nonNullValue);
+        } else if (key.equals(PATH_ORGANISATION.getNvaKey())) {
+            withPathOrganization(nonNullValue);
+        } else if (key.equals(PATH_PROJECT.getNvaKey()) || key.equals(PATH_PROJECT.getKey())) {
+            withPathProject(nonNullValue);
+        } else {
+            System.out.printf("CQB::SETPATH::INVALID_KEY -> [%s]-[%s]\n\r", key, value);
+            invalidKeys.add(key);
+        }
+    }
+
+    private void setValue(String key, String value) {
+        var qpKey = keyFromString(key,value);
+        if (!key.equals(qpKey.getKey()) && !cristinQuery.isNvaQuery && qpKey != INVALID) {
+            cristinQuery.isNvaQuery = true;
+        }
+        switch (qpKey) {
+            case PATH_ORGANISATION:
+                withPathOrganization(value);
+                break;
+            case PATH_PROJECT:
+            case IDENTITY:
+                withPathIdentity(value);
+                break;
+            case PROJECT_ORGANIZATION:
+                withOrganization(value);
+                break;
+            case BIOBANK:
+            case FUNDING:
+            case FUNDING_SOURCE:
+            case GRANT_ID:
+            case INSTITUTION:
+            case LEVELS:
+            case NAME:
+            case PAGE_CURRENT:
+            case PAGE_ITEMS_PER_PAGE:
+            case PAGE_SORT:
+            case PROJECT_APPROVAL_REFERENCE_ID:
+            case PROJECT_APPROVED_BY:
+            case PROJECT_KEYWORD:
+            case PROJECT_MANAGER:
+            case PROJECT_MODIFIED_SINCE:
+            case PROJECT_PARTICIPANT:
+            case PROJECT_UNIT:
+            case USER:
+                cristinQuery.setValue(qpKey, value);
+                break;
+            case TITLE:
+                withTitle(value);
+                break;
+            case LANGUAGE:
+                withLanguage(value);
+                break;
+            case QUERY:
+                withQuery(value);
+                break;
+            case STATUS:
+                withStatus(value);
+                break;
+            default:
+                System.out.printf("CQB::SETVALUE::INVALID_KEY -> [%s]-[%s]-[%s]\n\r", qpKey.name(), key, value);
+                invalidKeys.add(key);
+                break;
+        }
+    }
+
+    private boolean oneOrMoreOptionalIsMissing() {
+        return oneOrMoreOptionalUnassigned().size() == oneOrMoreOptional().size() && !requiredMissing().isEmpty();
+    }
+
+    private boolean invalidQueryParameter(QueryParameterKey key, String value) {
+        return isNull(value) || !value.matches(key.getPattern());
+    }
+
+    private boolean invalidPathParameter(QueryParameterKey key, String value) {
+        return !(value.isBlank() || value.matches(key.getPattern()));
```

Comment:
value.isBlank, should that be without the "!" ?

Reply:
if you have an easier way to express this, I am willing to listen :-)
path variables can be (with and valid identity or empty) else is invalid
If you take a look at where it is applied, it makes more sense ?

```
    private void throwInvalidPathValue(QueryParameterKey key, String value) throws BadRequestException {
        if (invalidPathParameter(key, value)) {
            throw new BadRequestException(errorMessage);
        }
    }
```